### PR TITLE
Add the Stream trait

### PR DESCRIPTION
**Pull request by sfackler**
This is a direct copy of the futures crate's version, which should enable that crate to reexport std's version in the future while preserving backwards compatibility. Like Future, this doesn't include
any of the various convenience methods - those can be added later and for now pulled from crates like futures.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
I'm going to r? @nikomatsakis here, though that's probably not quite the right person :)

Certainly not me, though.

File: src/libcore/stream.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    /// * `Poll::Pending` means that this stream's next value is not ready yet. Implementations will ensure that the
+    ///     current task will be notified when the next value may be ready.
+    /// * `Poll::Ready(Some(val))` means that the stream has successfully produced a value, `val`, and may produce
+    ///     further values on subsequent `poll_next` calls.
+    /// * `Poll::Ready(None)` means that the stream has terminated, and `poll_next` should not be invoked again.
+    ///
+    /// # Panics
+    ///
+    /// Once a stream is finished, i.e. `Ready(None)` has been returned, further calls may result in a panic or other
+    /// "bad behavior".
```
Comment (taiki-e) [inline src/libcore/stream.rs:37]:
```suggestion
    /// Once a stream has finished (returned `Ready(None)` from `poll_next`), calling its
    /// `poll_next` method again may panic, block forever, or cause other kinds of
    /// problems; the `Stream` trait places no requirements on the effects of
    /// such a call. However, as the `poll_next` method is not marked `unsafe`,
    /// Rust's usual rules apply: calls must never cause undefined behavior
    /// (memory corruption, incorrect use of `unsafe` functions, or the like),
    /// regardless of the stream's state.
```

As with [`Future`'s documentation](https://docs.rs/futures/0.3.5/futures/prelude/trait.Future.html#panics), it needs to be mentioned that "bad behavior" does not contain UB.

**Comment by nikomatsakis** [comment]
From a technical perpsective, 👍.

I guess the main question is whether we ought to have an RFC. I think I'd appreciate a bit of background on the design. But I know that the libs team tends to operate more with PRs than RFCs.

That said, we were recently [reviewing some of the known concerns](https://rust-lang.github.io/wg-async-foundations/design_notes/stream.html) with the stream trait and I think there weren't really any *blockers* -- 

* it uses a poll-based approach, which is required to support `dyn Stream`, but makes manual impls more difficult
* it doesn't permit "attached streams", which re-use internal resources, but then that's an orthogonal problem that will require GATs and experimentation to solve, and doesn't seem worth blocking on (also true of `Iterator`)
* one question was whether we should be added any add'l combinators off the bat (but then, those can likely be added later -- although that can often break code in practice if not in theory -- and many of those combinators either inhibit no-std compatibility or may be better with async-closures)

**Comment by sfackler** [comment]
> * one question was whether we should be added any add'l combinators off the bat

We've historically had troubles pulling in Iterator methods from things like itertools, since a call to `some_iterator.foo()` suddenly becomes ambiguous when `Iterator::foo` exists along side `itertools::IteratorExt::foo`. The same thing would happen here when pulling in anything from `futures::StreamExt` or `TryStreamExt`. We definitely need to be able to do this, though! One thing we've discussed is some (probably perma-unstable) way of telling the compiler that a trait's methods should be considered "low priority" when trying to resolve conflicts. That way code that uses `StreamExt` continues to behave the same when we start pulling methods in.

**Comment by nikomatsakis** [comment]
@sfackler yep, I had these troubles in mind. We would I think also want to add combinators to `Future` and a number of other traits as well.

**Comment by sfackler** [comment]
I do think we'll definitely want async generator syntax to make impl Streams, but I don't think that we need to block the addition of the trait on having it be "special" in that way.

**Comment by nikomatsakis** [comment]
So I started a [draft RFC](https://github.com/rust-lang/wg-async-foundations/blob/master/rfc-drafts/stream.md) -- should we block on completing that before landing this PR?

**Comment by Dylan-DPC** [comment]
this is blocked on a rfc which will take some time to get merged. I'm closing this to avoid it from rotting further. I think it would be easier to open a new PR once the rfc is merged.

**Pull request closed without merging**